### ISSUE TITLE
Automatic closing of HTML tags messes up inner text

### ISSUE THREAD

**Issue by YorickPeterse** [YorickPeterse/oga#104]
Per @abotalov:

> One more example:
> ```html
>  <object>
>   <ul>
>    <li>Li
>   </ul>
>   Inside object
>  </object>
>  Outside object
> ```
> is currently equivalent to:
> ```html
>  <object>
>   <ul>
>    <li>Li
>   </li>
>   Inside object
>  </ul>
>  Outside object
> </object>
> ```

**Comment by YorickPeterse** [comment]
This is due to the lexer treating this input as following:

```html
<object>
    <ul>
        <li>Li</li>
    Inside object
</ul>
Outside object
```

**Comment by YorickPeterse** [comment]
I have a fix for this on my computer, but it requires exposing element names to `XML::Lexer#on_element_end`. Because this will decrease performance a bit (only for HTML documents) I'll be giving this some extra thought to see how bad it is and if there's maybe not a way around this.

**Comment by YorickPeterse** [comment]
Fixed by 5182d0c488759efb96d85a399de29550faea3efe.

**Comment by abotalov** [comment]
It looks to me that the spec `lexes nested <ul> elements containing unclosed <li> elements` is incorrect.
Currently 
```html
<ul><li><ul><li>foo</ul><li>bar</ul>
```
is equivalent to:
```html
<ul><li><ul><li>foo</li><li>bar</li>
</ul></li></ul>
```

`<li>bar</li>` shouldn't be a child of a nested `ul` element.
